README.md

```diff
@@ -211,42 +212,163 @@ val records = sc.newAPIHadoopFile(
+
+The following describes how each connection can be authenticated:
+
+- **Spark driver to Redshift**: The Spark driver connects to Redshift via JDBC using a username and password.
+    Redshift does not support the use of IAM roles to authenticate this connection.
+    This connection can be secured using SSL; for more details, see the Encryption section below.
+
+- **Spark to S3**: S3 acts as a middleman to store bulk data when reading from or writing to Redshift.
+    Spark connects to S3 using both the Hadoop FileSystem interfaces and directly using the Amazon
+    Java SDK's S3 client.
+
+    This connection can be authenticated using either AWS keys or IAM roles (DBFS mountpoints are
+    not currently supported, so Databricks users who do not want to rely on AWS keys should use
+    cluster IAM roles instead).
+
+    There are multiple ways of providing these credentials:
+
+    1. **Default Credential Provider Chain (best option for most users):**
+        AWS credentials will automatically be retrieved through the [DefaultAWSCredentialsProviderChain](http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#id6).
+
+        If you use IAM instance roles to authenticate to S3 (e.g. on Databricks, EMR, or EC2), then
+        you should probably use this method.
+
+        If another method of providing credentials is used (methods 2 or 3), then that will take
+        precedence over this default.
+
+    2. **Set keys in Hadoop conf:** You can specify AWS keys via
+        [Hadoop configuration properties](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md).
+        For example, if your `tempdir` configuration points to a `s3n://` filesystem then you can
+        set the `fs.s3n.awsAccessKeyId` and `fs.s3n.awsSecretAccessKey` properties in a Hadoop XML
+        configuration file or call `sc.hadoopConfiguration.set()` to mutate Spark's global Hadoop
+        configuration.
+
+        For example, if you are using the `s3n` filesystem then add
+
+        ```scala
+        sc.hadoopConfiguration.set("fs.s3n.awsAccessKeyId", "YOUR_KEY_ID")
+        sc.hadoopConfiguration.set("fs.s3n.awsSecretAccessKey", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+        and for the `s3a` filesystem add
+
+        ```scala
+        sc.hadoopConfiguration.set("fs.s3a.access.key", "YOUR_KEY_ID")
+        sc.hadoopConfiguration.set("fs.s3a.secret.key", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+        Python users will have to use a slightly different method to modify the `hadoopConfiguration`,
+        since this field is not exposed in all versions of PySpark. Although the following command
+        relies on some Spark internals, it should work with all PySpark versions and is unlikely to
+        break or change in the future:
+
+        ```python
+        sc._jsc.hadoopConfiguration().set("fs.s3n.awsAccessKeyId", "YOUR_KEY_ID")
```

Comment:
Is there a non-internal method that may not be as pretty or is this the best we can do?

i.e. for pyspark > N use this supported method, for pyspark <= N use this workaround